Add UPP libraries

mkavulich opened the pull request:
I am aware that there are moves to have an official NCEPlibs umbrella build that will be released. But with the timeline and portability of that effort uncertain, and this work completed anyway, I'd like to merge in the build capabilities for these new libraries, which allows UPP to build with these as an externally built package.

This pull request adds the following NCEP libraries to this framework:
 - g2
 - g2tmpl
 - gfsio
 - wrfio

For these new libraries, if the user desires them, they must set some additional environment variables pointing to other external libraries:

 - jasper
 - png
 - netcdf

These libraries have been tested on the following platforms:

 - Macosx
   - Gnu
     - 4.9.4
     - 5.5.0
 - Yellowstone
   - Gnu 8.3.0 (default)
   - Intel 18.0.5 (default)
   - PGI 19.3 (default)
 - Hera
   - Intel (default)
 - Generic Linux
   - gnu 4.9.2

There is at least one known gnu version for which g2 libraries fail to build: gnu 9.2.0. This will require a bug fix in the library code itself. I believe this bug fix is already in the works so I have left it alone for now.

I do not have access to other machine/compiler combinations for testing.

A few other improvements have been bundled into this pull request as well:

 - If user sets environment variable NOCOMPILERCHOICE, this will skip the manual diolog for confirming compilers on generic linux and macosx platforms
 - Library names will be built with their full names (e.g. libsigio_v2.0.1_4.a) but also linked with their generic shortened names (without version, realsize, etc info, e.g. libsigio.a). This is done with some bash string editing that admittedly has a chance of failure if new libraries are introduced with weird naming conventions. But the logic will never overwrite valid files.

mkavulich committed:
Adding initial version of g2 libraries (v3.1.0) and initial bits for compiling upp library set

mkavulich committed:
A couple of minor text fixes

mkavulich committed:
Removing old line causing compilation failures with newer versions of jasper lib. More information here: https://github.com/wrf-model/WPS/pull/50

mkavulich committed:
g2 library builds for gnu on Cheyenne

mkavulich committed:
Add initial version of g2tmpl libraries (1.5.0)

mkavulich committed:
g2tmpl library builds for gnu on Cheyenne

mkavulich committed:
Fix top-level 'make clean' functionality

mkavulich committed:
Add initial version of gfsio library (1.1.0)

mkavulich committed:
gfsio library builds for gnu on Cheyenne

mkavulich committed:
Add initial version of wrfio library (1.1.1)

mkavulich committed:
Remove nemsio makerule that deletes .o files; this rule forces make to rebuild the libraries even after successful compilation

mkavulich committed:
wrfio library builds for gnu on Cheyenne

mkavulich committed:
Somehow forgot to add g2tmpl makefile before now

mkavulich committed:
Include links to "simplified" names of library files; for example, link libg2_v3.1.0_4.a to simplified name libg2.a.

mkavulich committed:
When building, copy .mod files for g2 and g2tmpl to include directory

mkavulich committed:
Fix g2 lib flags for proper LINUX build

mkavulich committed:
A few more modifications to get everything linked together properly for UPP

 - gfsio actually depends on w3emc; should note this and ensure everything is included properly
 - Using the correct gfortran compile flags for gfsio and w3 libraries
 - Copy the .mod files from w3 and gfsio into the include directory at compile time

mkavulich committed:
UPP requires the params_grib2_tbl_new template file from g2tmpl library, so we should copy it to the include directory at compile time

mkavulich committed:
Add make rules for UPP libraries for Cheyenne, intel

mkavulich committed:
New libraries compile successfully for Intel on Hera

mkavulich committed:
Include ability to force past the command line choice of compilers for macosx and linux using environment variable NOCOMPILERCHOICE

mkavulich committed:
All libraries work for generic linux now, gnu only so far (as I am unable to test other compilers)

mkavulich committed:
Compilation successful for Cheyenne, PGI

mkavulich committed:
Update macros file for for macosx, gnu for UPP libraries

mkavulich committed:
Openmp must be disabled for UPP library compilation

mkavulich committed:
UPP libraries *are* ready!

mkavulich committed:
A few fixes for Cheyenne, pgi macros file

climbfuji commented:
This is great work, thanks very much. I am actually adding GNU portability to the new NOAA-EMC NCEPlibs umbrella build right now, which builds everything here and more (grib utilities, NCEP-post, ...).

Nonetheless, this could be really useful. I noticed a few things:
- compiler option changes for existing libraries: I need to check whether this affects any of the ufs-weather-model regression test results
- OpenMP: why is OpenMP not possible when "upp" is selected?
- Linux: why is Linux not available when "upp" is selected? Just because of no test system available?
- Preprocessor macros: for one of the existing libraries (BACIO), I modified the code to include #ifdef MACOSX in addition to/separately from LINUX. I think we should either do the same for all others or (my preferred choice) see if we can remove the MACOSX preprocessor macros that I defined in BACIO and use LINUX instead.

mkavulich committed:
Various top-level fixes

 - Check for upp-required environment variables if $APP is 'all', not just 'upp'
 - Remove a stray debug echo statement
 - Fix comments in Makefile to properly state that iplib is NOT needed for UPP;
   since this was a core library from the beginning will not try to change makefile
   rules or priorities

mkavulich commented:
1. Regarding the changes to the macros files for existing libraries:
   - I changed the flags for w3 libraries in macros.make.cheyenne.gnu during some testing I was doing, but this was not a necessary change and I am comfortable reverting it if necessary
   - the new IPlib variable in macros.make.linux.gnu was necessary for building on one of the generic linux platforms I tested, I could not get it to build otherwise even in the existing master branch. This mirrors a similar variable already found in other linux gnu macros files
2. This was put in specifically a restriction for UPP itself, not the libraries per se (since UPP does not support an openmp build). However, it appears as if this was probably not necessary; I just built UPP with the openmp version of the libraries and it appears to work fine; I'll do a few more tests but I think I can remove that restriction.
3. I think you're misinterpreting the change on line 130 of make_ncep_libs.sh. Generic linux *is* now available for UPP and SAR, previously it was *not* available for SAR. I actually don't anticipate any restrictions on the platforms available for the new libraries, but I have only added macros files for those platforms that I have access to.
4. I didn't define any new macros in the new libraries, but I did make use of some existing #LINUX macros in some of my mods. I'm open to any suggested changes that you have regarding this.

climbfuji commented:
Great, thank you! I suggest you merge this after trying the OpenMP build with UPP, and I will create an issue to clean up the preprocessor arguments later on. Greetings from Japan!

mkavulich commented:
@climbfuji I have discovered an issue related to UPP's interaction with some NCEP libs that might be a problem. I previously said that the flag changes for w3 were not necessary and could be reverted. Well it turns out that not only are they necessary, but I need to implement them for Intel and PGI as well. The reason is that UPP builds with 4-byte reals by default, and when attempting to use the w3nco lib compiled with 8-byte reals, this results in pernicious errors due to UPP interpreting arrays of 8-byte reals as arrays of 4-byte reals. I didn't even think this condition was possible, but it was resulting in errors I've spent the past week tracking down.

Do you think changing w3nco to use 4-byte reals will be a problem? If it might be an issue, I could also make these compilation flags an option specific to UPP, though it would likely increase the build complexity a bit.

climbfuji commented:
Thanks for the update. I think that needs to be UPP-specific. There is a reason why it had to be in double precision for the UFS weather model. But this makes it difficult, one would have to compile two different sets of NCEPlibs, one for UPP and one for UFS. How about compiling two versions, single and double precision, one after the other, and copying the files with suffix _4 and _8 (this is usually done for NCEPlibs)?

mkavulich commented:
> How about compiling two versions, single and double precision, one after the other, and copying the files with suffix _4 and _8 (this is usually done for NCEPlibs)?

Good suggestion, this may be a better idea. I will work on this and update the branch accordingly.

climbfuji commented:
Let me note that I made progress on EMC's new NCEPLIBS umbrella build system for macOS/GNU. This builds the following libraries:
```
libbacio_v2.1.0_4.a
libbacio_v2.1.0_8.a
libbufr_v11.3.0_4.a
libbufr_v11.3.0_8.a
libbufr_v11.3.0_d.a
libcrtm_v2.3.0.a
libg2_v3.1.0_4.a
libg2_v3.1.0_d.a
libg2tmpl_v1.5.0_d.a
libgfsio_v1.1.0_4.a
libip_v3.0.0_4.a
libip_v3.0.0_8.a
libip_v3.0.0_d.a
liblandsfcutil_v2.1.0_4.a
liblandsfcutil_v2.1.0_d.a
libnemsio_v2.2.3.a
libnemsiogfs_v2.2.1.a
libsfcio_v1.1.0.a
libsigio_v2.1.0.a
libsp_v2.0.2_4.a
libsp_v2.0.2_8.a
libsp_v2.0.2_d.a
libw3emc_v2.2.0_4.a
libw3emc_v2.2.0_8.a
libw3emc_v2.2.0_d.a
libw3nco_v2.0.6_4.a
libw3nco_v2.0.6_8.a
libw3nco_v2.0.6_d.a
```
alongside with the corresponding `.mod` Fortran include files. It also builds the following binaries:
```
cnvgrib
copygb
copygb2
degrib2
fsync_file
grb2index
grbindex
grib2grib
mdate
ncep_post
ndate
nhour
tocgrib
tocgrib2
tocgrib2super
wgrib
```

mkavulich committed:
UPP needs 4-byte reals, so we need to build both 4-byte and 8-byte real versions of w3 libraries. Additionally, will no longer attempt to link "simplified" versions of the libraries, since we need to keep the real-size suffixes intact

NOTE: this may result in changed results; previously "libw3emc_d" was NOT actually double-precision (8-byte reals). Now it is correctly compiled that way.

climbfuji commented on src/w3emc/v2.2.0/src/makefile_d:
```diff
@@ -4,7 +4,7 @@ include ../../../../macros.make
 version   = v2.2.0
 tarname   = w3emc
-precision = d
+precision = 4
```
Why is the precision changed from "d" to "4" in makefile_d?

mkavulich replied:
Good catch! That was definitely an error. I had some fumbling when I was trying to negotiate the different naming conventions for all the different libraries; originally I started out trying to standardize all the library names but then I thought that would be too drastic of a change for this pull request.

As I mentioned in the commit message, I noticed something weird about the previous libw3emc build. Even though the library was built with the name "libw3emc_d.a", this was actually built with single/4-byte precision reals, at least for the Cheyenne macros files. I have changed it so that the expected precision corresponds to the appropriate filename, as now both precisions are built. This may have some impact on UFS results however, since it (perhaps unexpectedly) was previously using that library compiled for single precision.

If you think this was an inappropriate change, please let me know.

mkavulich committed:
Fix erroneous change in src/w3emc/v2.2.0/src/makefile_d; obviously precision should be "d"

mkavulich committed:
Need to make clean in between building libraries for different precisions

mkavulich committed:
w3emc lib needs -c flag for gnu compilers

mkavulich committed:
Add new flags for 4 and 8 byte w3 libraries on MacOSX

climbfuji commented on macros.make.macosx.gnu:
```diff
@@ -4,9 +4,10 @@
-#FCserial   = gfortran
-#CC         = gcc
+FC         = mpif90
+FCserial   = gfortran-5
```
Please don't change the defaults here, on my Mac those would need to be gfortran-9 and gcc-9. Instead, use the prompt during the installation to change the default values.

mkavulich replied:
Fixed, sorry this was not intended to be committed

mkavulich committed:
Some W3 compile flag fixes for Hera

mkavulich commented:
@climbfuji I believe all the requested changes have been included, and the changes should now allow UPP libraries to compile for all the above-mentioned platforms. Let me know if there's any additional tests you'd like me to run.

mkavulich committed:
Undoing unintentional changes to macosx default compilers

climbfuji commented:
Hey Mike, I am sorry that this is taking so long. I just checked out your branch and compiled it, that all works. But I can't compile the UFS weather model on my mac (or any other machine that is using the NCEPLIBS from this repository), because the names of the libraries have changed. If we commit your PR, we need to update some of the modulefiles in the ufs-weather-model repo or on the systems (cheyenne, jet, gaea, ...).

If we have to do that, we (i.e. us) might as well move to the new NCEPLIBS from the umbrella build I suppose, and then hand off this entire repository to you no longer need it.

mkavulich commented:
Dom, Are there any projects that are still using this version of NCEPlibs aside from the UPP stand-alone build? I was under the impression that the ufs-weather-model repo was going to use the official umbrella NCEPLIBS build. 

I can revert the library name changes but that would make them inconsistent with the traditional EMC naming conventions. Honestly I'm okay with whatever is easiest for you or anyone else that is still using this repository; I'd like to get this finished up ASAP.

climbfuji commented:
> Dom, Are there any projects that are still using this version of NCEPlibs aside from the UPP stand-alone build? I was under the impression that the ufs-weather-model repo was going to use the official umbrella NCEPLIBS build.
> 
> I can revert the library name changes but that would make them inconsistent with the traditional EMC naming conventions. Honestly I'm okay with whatever is easiest for you or anyone else that is still using this repository; I'd like to get this finished up ASAP.

Several systems use this for the ufs-weather-model (generic macOS, Linux, Cheyenne, Jet, Gaea, ...). If you could create a tag of the current head with a meaningful name (e.g. nceplibs-ufs-weather-model-20200220) and description (if there is such a feature), then I am happy for you to go ahead.
